pydantic/_internal/_schema_gen/_generate_schema.py

```diff
@@ -0,0 +1,365 @@
+        self._typevars_map = typevars_map
+        self.field_name_stack = _FieldNameStack()
+        self.model_type_stack = _ModelTypeStack()
+        self.defs = _Definitions()
+
+    def generate_schema(self, obj: Any) -> CoreSchema:
+        # 1. Check GPCS method
+        schema = self._generate_schema_from_get_schema_method(obj)
+
+        if schema is None:
+            schema = self._generate_schema_inner(obj)
+
+
+        metadata_js_function = _extract_get_pydantic_json_schema(obj, origin)
+        if metadata_js_function is not None:
+            ...
+
+        return schema
+
+    def _generate_schema_inner(self, obj: Any) -> CoreSchema:
+        if typing_objects.is_self(obj):
+            ...
+
+        origin = get_origin(obj)
+
+        if typing_objects.is_annotated(origin):
+            return self._generate_schema_with_metadata(
+                obj=obj.__origin__,  # <typ> in Annotated[<typ>, 'meta']
+                metadata=obj.__metadata__,  # ('meta',) in Annotated[<typ>, 'meta']
+            )
+
+        schema = self._match_type(
+            obj,
+            _origin=origin,
+        )
+
+        return schema
+
+    def _match_type(
+        self,
+        obj: Any,
+        _origin: Any = PydanticUndefined,
+    ) -> CoreSchema:
+        origin = _origin if _origin is not PydanticUndefined else get_origin(obj)
+
+        if is_union_origin(origin):
+            return self._union_schema(obj)
+
+        if typing_objects.is_literal(origin):
+            return self._literal_schema(obj)
+
+        if origin is not None and (aliased_obj := typing_objects.DEPRECATED_ALIASES.get(obj)):
+            obj = aliased_obj
+            origin = None
+
+        type_handler_class = self._type_registry.get_type_handler(origin if origin is not None else obj)
+        if type_handler_class is None:
+            raise Exception(f'No handler for {origin if origin is not None else obj}')
+
+        type_handler = type_handler_class(self)
+
+        ref = type_handler.get_reference(origin, obj)
+        if ref is not None:
+            with self.defs.get_schema_or_ref(ref) as maybe_schema:
+                if maybe_schema is not None:
+                    return maybe_schema
+
+                core_schema = type_handler._generate_schema(origin, obj)
+                return core_schema
+        else:
+            core_schema = type_handler._generate_schema(origin, obj)
+            return core_schema
+
+    def _literal_schema(self, literal_type: Any) -> CoreSchema:
+        expected = list(get_literal_values(literal_type, type_check=False, unpack_type_aliases='eager'))
+        schema = core_schema.literal_schema(expected)
+
+        # if self._config_wrapper.use_enum_values and any(isinstance(v, Enum) for v in expected):
+        #     schema = core_schema.no_info_after_validator_function(
+        #         lambda v: v.value if isinstance(v, Enum) else v, schema
+        #     )
```

Comment:
If this is necessary, should it go in core to avoid the Python overhead / wrapper?